consensus/istanbul/utils.go

```diff
@@ -67,72 +67,56 @@ func CheckValidatorSignature(valSet ValidatorSet, data []byte, sig []byte) (comm
 	return common.Address{}, fmt.Errorf("not an elected validator %s", signer.Hex())
 }
 
-// Retrieves the block number within an epoch.  The return value will be 1-based.
-// There is a special case if the number == 0.  It is basically the last block of the 0th epoch, and should have a value of epochSize
+// GetNumberWithinEpoch retrieves the block number within an epoch.
+// The return value will be 1-based; thus first block of epoch is 1, and last block of epoch is `epochSize`
+// There is a special case if the number == 0. It is basically the last block of the 0th epoch, and should have a value of epochSize
 func GetNumberWithinEpoch(number uint64, epochSize uint64) uint64 {
 	number = number % epochSize
 	if number == 0 {
 		return epochSize
-	} else {
-		return number
 	}
+	return number
 }
 
+// IsLastBlockOfEpoch indicates if block number is the last block of its epoch
 func IsLastBlockOfEpoch(number uint64, epochSize uint64) bool {
 	return GetNumberWithinEpoch(number, epochSize) == epochSize
 }
 
+// IsFirstBlockOfEpoch indicates if block number is the first block of its epoch
 func IsFirstBlockOfEpoch(number uint64, epochSize uint64) bool {
 	return GetNumberWithinEpoch(number, epochSize) == 1
 }
 
-// Retrieves the epoch number given the block number.
-// There is a special case if the number == 0 (the genesis block).  That block will be in the
-// 1st epoch.
+// GetEpochNumber retrieves the epoch number given the block number.
+// Epoch 0 is a special block that only contains the genesis block (block 0), epoch 1
+// starts at block 1
 func GetEpochNumber(number uint64, epochSize uint64) uint64 {
 	epochNumber := number / epochSize
```

Comment:
Maybe rename `epochNumber` to `epochs` or `pastEpochs` or something else.  It just seems strange to have a function `GetEpochNumber()` with a return that says `epochNumber + 1`

Reply:
Not sure i agree...
`epochNumber` is the result variable; so it's proper to named as this.

What could clarify this is do:
```
epochNumber = epochNumber +1
return epochNumber
```

Reply:
That would be better.  We could also name the return variable:
```
func GetEpochNumber(number uint64, epochSize uint64) (epochNumber uint64) {
	epochNumber = number / epochSize
	if !IsLastBlockOfEpoch(number, epochSize) {
		epochNumber += 1
	}
}
```

There are other ways to make the special case of the last block more explicit (including with comments), but most of them don't necessarily make the code better in the end.  I almost think the most self-explanatory and readable form is when you duplicate the division:

```
func GetEpochNumber(number uint64, epochSize uint64) uint64 {
	if IsLastBlockOfEpoch(number, epochSize) {
		return number / epochSize
	} else {
		return number / epochSize + 1
	}
}
```

Reply:
But probably this is just hair-splitting...

Reply:
went with the last version proposed....
totally hair-splitting :P